Fix `Size` in `docker images` for v1.12 client to v1.13 daemon

yongtang opened the pull request:
**- What I did**

This fix tries to address the issue raised in #30027 where `Size` in `docker images` always show `-1` for v1.12 client against v1.13 daemon.

The reason for the issue is that in v1.12, `Size` is used while in v1.13 `VirtualSize` is used. And in v1.13, `Size` has been always initialized as `-1`.

**- How I did it**

This fix address the issue by copy the value of `VirtualSize` to `Size` if version is less or equal to v1.24 (docker v1.12).

**- How to verify it**

An integration test has been added.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

![u 1024162214 1388002003 fm 21 gp 0](https://cloud.githubusercontent.com/assets/6932348/21853212/ff0f1736-d7ca-11e6-9229-27ac405bfbc9.jpg)


This fix fixes #30027.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

AkihiroSuda commented on integration-cli/docker_api_images_test.go:
```diff
@@ -128,3 +128,38 @@ func (s *DockerSuite) TestAPIImagesSearchJSONContentType(c *check.C) {
 	c.Assert(res.Header.Get("Content-Type"), checker.Equals, "application/json")
 }
+
+// Test case for 30027
```
nit: please add a brief description about #30027? (e.g. "Image size reported as -1 in v1.12 client against v1.13 daemon")

yongtang replied:
Thanks @AkihiroSuda. The PR has been updated.

thaJeztah commented:
ping @mlaventure

mlaventure committed:
Fix ImageSummary.Size value

The prune PR changed the meaning of the file to mean "space on disk
only unique to this image", this PR revert this change.

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

mlaventure commented:
@yongtang we decided to fix the meaning of the field instead, but I feel like the PR should bear your name.

Do you mean using this commit: https://github.com/mlaventure/docker/commit/d143fb8bee428074c91ec901e204e23b318c3cc3 

And add your test to it?

Thanks!

yongtang committed:
Add test for image size for v1.12 and v1.13 clients against v1.13 daemon

This test checks to make sure both v1.12 and v1.13 client against v1.13 daemon get correct `Size` after the fix.

This test is related to 30027.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@mlaventure Thanks for the help! The commit has been carried and test has been added. Please take a look and let me know if there are any issues.

thaJeztah reviewed (approved):
LGTM

mlaventure reviewed (approved):
LGTM

Thanks @yongtang !

ZYNCMA commented:
> - How I did it
>
>This fix address the issue by copy the value of VirtualSize to Size if version is less or equal to v1.24 (docker v1.12).

@yongtang Did you really judge the api version before assigning virtualsize to size?
 I cannot find it in your commits. Did I miss anything?

thaJeztah commented:
@ZYNCMA `Size` and `VirtualSize` were used before Docker 1.10 (which introduced content addressable store). Before 1.10, `Size` was the size of just the last layer of the image, and `VirtualSize` the calculated size of the image + all its parents. After Docker 1.10, images are self-contained, so `Size` and `VirtualSize` will always be the same.

This PR reverts a change that re-used `Size`, and set it to_just_ the size of layers that are not shared with other images. We decided it was bad to change the definition of that field, so reverted the change; therefore this change does not have to take the API version into account.

You can see the original change here; https://github.com/docker/docker/pull/26108/commits/b717de5153db503fae61c11e30f04f400f66a6fd#diff-4adce79060ec85cddf1545ebadcc0d5eR244

(I see we didn't change back the _name_ of the parameter; let me open a separate PR for that, but it's not critical, just a "nit")

thaJeztah commented:
Opened https://github.com/docker/docker/pull/30120 to change back the name

ZYNCMA commented:
@thaJeztah I'm fine with the PR. I was just confused about the inconsistency of the description I quoted before and the commit.
As it's unnecessary to take the API version into account, it should be the fault of the description.
Thanks for your explanation!

ZYNCMA commented:
@yongtang Got it.
